torch/nested/_internal/nested_tensor.py

```diff
@@ -285,5 +305,41 @@ def jagged_from_list(
     return ViewNestedFromBuffer.apply(values, offsets), offsets  # type: ignore[call-overload]
 
 
+def jagged_from_tensor_and_lengths(
+    tensor: torch.Tensor, starts: torch.Tensor, lengths: torch.Tensor
+) -> Tuple[NestedTensor, torch.Tensor, torch.Tensor]:
+    """Constructs a NestedTensor backed by jagged layout from a tensor, starts of sequences, and sequence lengths"""
+    batch_size = tensor.shape[0]
+    if is_expandable_to(starts.shape, (batch_size, 1)) and is_expandable_to(
+        lengths.shape, (batch_size, 1)
+    ):
+        start_list = starts.expand(batch_size, 1)
+        length_list = lengths.expand(batch_size, 1)
+    else:
+        raise RuntimeError(
+            "When constructing a jagged nested tensor using narrow(), "
+            "your start and length must be a Tensor that broadcasts to input.shape[0] x 1"
+        )
+
+    # Calculate jagged offsets
+    assert (
+        len(tensor.shape) >= 2
+    ), "tensor must at least be 2D for the nested narrow op to work"
+    max_seq_len = tensor.shape[1]
+    offset_lengths = max_seq_len * torch.arange(
+        0, batch_size, dtype=torch.int64, device=tensor.device
+    )
+    # Jagged layout specifies that offsets are stored as int64 on the same device as values.
+    offsets = start_list + offset_lengths
+
+    # Reshape buffer to flatten the 1st and 2nd dimension
+    if len(tensor.shape) > 2:
+        values = tensor.reshape(-1, *tensor.shape[2:])
```

Comment:
hmm this makes me think whether we should make assertions about the contiguity of the input tensor. 
both reshape and flatten can conditionally copy depending whether the input shape is cleanly reshapable to the desired shape. 
I feel like we at least do something like view to make users explicitly contiguify if necessary to avoid any silent performance cliffs when inputs change.

Reply:
> I feel like we at least do something like view to make users explicitly contiguify if necessary to avoid any silent performance cliffs when inputs change.

probably a good idea to require being explicit about this, I agree